### PR TITLE
Adding airflow operator to submit and monitor Spark Apps/App Env on DPGDC clusters

### PR DESCRIPTION
**Pull request by akashsriv07**
Adding airflow operator to create and monitor Spark Apps/App Env on DPGDC clusters
---

GDC extends Google Cloud’s infrastructure and services to customer edge locations and data centers. We intend to bring [Dataproc](https://cloud.google.com/dataproc?hl=en) as a managed service offering to GDC.

As part of this, we need to integrate Airflow with Dataproc on GDC API resources to trigger Apache Spark workloads. The integration needs to cover two distinct API paths: local execution through the KRM API and remote execution through an equivalent One Platform API. We're targeting the former one in this PR i.e. Airflow operator leveraging KRM APIs.

This PR contains two feature implementations :
1. An Airflow operator can submit a SparkApplication via the KRM API, enabling customers to use Airflow to orchestrate Dataproc on GDC workloads locally. 
2. An Airflow operator can create an ApplicationEnvironment via both the KRM API.

**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/operators/dataprocgdc.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+from kubernetes.client import CoreV1Api, ApiException
+from kubernetes.watch import Watch
+
+
+if TYPE_CHECKING:
+    from kubernetes.client.models import CoreV1EventList
+    from airflow.utils.context import Context
+
+
+class DataprocGDCSubmitSparkJobKrmOperator(BaseOperator):
```
Comment (dirrao) [inline airflow/providers/google/cloud/operators/dataprocgdc.py]: Can't we extend the functionality of existing operator `DataprocSubmitSparkJobOperator` ?

Comment (akashsriv07) [reply airflow/providers/google/cloud/operators/dataprocgdc.py]: There are two ways we need to go for implementing Airflow operator to trigger workloads on GDC cluster.
1 is via KRM API and another is via Google cloud API(where we could use the DataprocSubmitSparkJobOperator, however need to check if that's possible as well).

This PR targets the KRM APIs exposed in a GDC cluster. So this cannot leverage the existing operator.

**Review by hussein-awala — COMMENTED** [review]
I didn't find anything about this in the GCP documentation, could you please add the documentation link?

Is the CRD `dataprocgdc.cloud.google.com/v1alpha1` based on on `sparkoperator.k8s.io/v1beta2`? I'm asking because we already have two operators for the `spark-on-k8s-operator`, so maybe we can use one of them as a superclass to your operator to avoid code duplication and implementing everything from scratch.

**Comment by akashsriv07** [comment]
Hey Hussein,
Dataproc on GDC is yet to go GA and this is one of the critical features to add as part of the same. Hence we don't have a public doc on the same.
You can take a look at this blog post though: https://cloud.google.com/blog/products/infrastructure-modernization/google-distributed-cloud-new-ai-and-data-services

The CRD for DPGDC is completely different than the sparkoperator.k8s.io/v1beta2, which stops us from leveraging the same operator/sensor.
using KRM APIs is one of the ways to interact with a DPGDC cluster.
In future, we plan to add the operator similar to dataproc.py(ex: DataprocSubmitSparkJobOperator) which will leverage Google's internal One Platform API mechanism.

> I didn't find anything about this in the GCP documentation, could you please add the documentation link?
> 
> Is the CRD `dataprocgdc.cloud.google.com/v1alpha1` based on on `sparkoperator.k8s.io/v1beta2`? I'm asking because we already have two operators for the `spark-on-k8s-operator`, so maybe we can use one of them as a superclass to your operator to avoid code duplication and implementing everything from scratch.

**Comment by hussein-awala** [comment]
> Hey Hussein, Dataproc on GDC is yet to go GA and this is one of the critical features to add as part of the same. Hence we don't have a public doc on the same. You can take a look at this blog post though: https://cloud.google.com/blog/products/infrastructure-modernization/google-distributed-cloud-new-ai-and-data-services

It will be a bit complicated to review this PR without a doc, I will try to review the syntax and check if our conventions are respected. (cc: @eladkal could you take a look?)

> The CRD for DPGDC is completely different than the sparkoperator.k8s.io/v1beta2, which stops us from leveraging the same operator/sensor. using KRM APIs is one of the ways to interact with a DPGDC cluster. In future, we plan to add the operator similar to dataproc.py(ex: DataprocSubmitSparkJobOperator) which will leverage Google's internal One Platform API mechanism.

No problem with that, my goal was to reduce code duplication if possible, but since they are different, we can keep building your operator from scratch.

File: airflow/providers/google/cloud/operators/dataprocgdc.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        else:
+            body = self.application_file
+        name = body["metadata"]["name"]
+        namespace = self.namespace or self.hook.get_namespace()
+        self.hook.delete_custom_object(
+            group=self.api_group, version=self.api_version, plural=self.plural, namespace=namespace, name=name
+        )
+
+
+class DataprocGdcCreateAppEnvironmentKrmOperator(BaseOperator):
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/operators/dataprocgdc.py]: For me these operators `DataprocGDCSubmitSparkJobKrmOperator`  and `DataprocGdcCreateAppEnvironmentKrmOperator` look quite similar to `KubernetesCreateResourceOperator` operator(https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/operators/resource.py#L107). Is it possible to use `KubernetesCreateResourceOperator` for your purpose or maybe you can be inherited from this operator?

Also, in this PR(https://github.com/apache/airflow/pull/37616) I have added ability to create namespaced 'custom_resource_definition' object and create this object on GKE cluster.

File: airflow/providers/google/cloud/sensors/dataprocgdc.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+from airflow.exceptions import AirflowException, AirflowSkipException
+from airflow.providers.cncf.kubernetes.hooks.kubernetes import KubernetesHook
+from airflow.sensors.base import BaseSensorOperator
+
+if TYPE_CHECKING:
+    from airflow.utils.context import Context
+
+
+class DataprocGDCKrmSensor(BaseSensorOperator):
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/sensors/dataprocgdc.py]: Is it possible to use this `SparkKubernetesSensor` (https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py#L33) for this purpose or maybe you can adopt this sensor for GDC cluster?

File: tests/system/providers/google/cloud/dataprocgdc/example_dataprocgdc.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+from airflow import DAG
+from airflow.providers.google.cloud.operators.dataprocgdc import (
+    DataprocGdcCreateAppEnvironmentKrmOperator,
+    DataprocGDCSubmitSparkJobKrmOperator,
+)
+from airflow.providers.google.cloud.sensors.dataprocgdc import DataprocGDCKrmSensor
+
+DAG_ID = "example_dag_dpgdc-krm"
+
+with DAG(
```
Comment (MaksYermak) [inline tests/system/providers/google/cloud/dataprocgdc/example_dataprocgdc.py]: Did you check this DAG on your local env?

File: tests/system/providers/google/cloud/dataprocgdc/example_dataprocgdc.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    DAG_ID, schedule="@once", catchup=False, start_date=datetime(2024, 2, 5), tags=["example", "dataprocgdc"]
+) as dag:
+    submitSparkJobOperator = DataprocGDCSubmitSparkJobKrmOperator(
+        task_id="example-dataprocgdc-submitspark-operator",
+        trigger_rule="all_success",
+        depends_on_past=False,
+        retries=1,
+        application_file="sparkpi.yaml",
+        namespace="default",
+        kubernetes_conn_id="myk8s",
```
Comment (MaksYermak) [inline tests/system/providers/google/cloud/dataprocgdc/example_dataprocgdc.py]: Does this operator work with GKE cluster configuration?

File: airflow/providers/google/cloud/sensors/dataprocgdc.py (hunk 3 above)
Comment (akashsriv07) [reply airflow/providers/google/cloud/sensors/dataprocgdc.py]: I thought around that, however the response states in GDC is different from the ones in spark Kubernetes application mentioned by you.

File: tests/system/providers/google/cloud/dataprocgdc/example_dataprocgdc.py (hunk 4 above)
Comment (akashsriv07) [reply tests/system/providers/google/cloud/dataprocgdc/example_dataprocgdc.py]: Yes.

**Comment by akashsriv07** [comment]
> Hi @akashsriv07,
> 
> I will propose a different approach for the folder/module structure. Since GDC might have different APIs, or different inner working mechanisms, it can be good to separate the GDC related things (like `hooks`, `operators`, `sensors`) from the `cloud` one. In the future, if you want to add more things (like `hooks`, `operators`, `sensors`), it can be easily extended in a more structured way. Also, you will have access for the existing Google Cloud `operators`, `sensors`, etc.
> 
> Hence, I am suggesting creating a new module inside a `google` provider (at the same level with the `cloud` module). Something like the following:
> 
> ```
> airflow/providers/google
> ├── ads
> ├── cloud
> ├── common
> ├── firebase
> ├── gdc                         # <-- this is the new one
> ├── leveldb
> ├── marketing_platform
> ├── suite
> ├── CHANGELOG.rst
> ├── go_module_utils.py
> ├── __init__.py
> └── provider.yaml
> ```
> 
> What do you think?

Hey @molcay , GDC is part of Google Cloud only. Shouldn't it be a part of it?

**Comment by molcay** [comment]
Hey @akashsriv07,

When I read the following sentence on [the product page](https://cloud.google.com/distributed-cloud#section-1), I thought that this product **is not directly in the GCP**, rather it is equvienlt to the GCP but it is for on-premise infrastructure:
> Build, deploy, and scale modern industry & public sector applications on-premise with AI ready modern infrastructure, ensure data security, and enable agile developer workflows across edge locations & data centers.

But of course, I am not entirely sure on the topic. This suggestion folds on a grey area I guess.
But at the base level, maybe we can group the changes related to GDC in a single module and put it to inside of the cloud module.

```bash
airflow/providers/google/cloud
├── example_dags
├── fs
├── gdc                         # <-- this is the new one
├── hooks
├── __init__.py
├── _internal_client
├── links
├── log
├── operators
├── __pycache__
├── secrets
├── sensors
├── transfers
├── triggers
└── utils
```

Also, maybe it is too early to have this module structure right now. We can also wait and see how it will extend

**Comment by michalmodras** [comment]
Hey, Michal from Cloud Composer team here. 
+1 for this code living in a separate, GDC provider - GDC is based on slightly different infra than GCP, and the operators won't necessarily work in the same way (some might be very similar, some very different). If there's shared code, I propose having a util module/provider between these two. 
FWIW we plan to separate out more Google providers (e.g. taking out ads to a separate one) to decouple support of Google services a bit, instead of having them all in one provider, at times unnecessarily tangling e.g. dependencies.

File: tests/system/providers/google/gdc/dataprocgdc/example_dataprocgdc.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    DAG_ID, schedule="@once", catchup=False, start_date=datetime(2024, 2, 5), tags=["example", "dataprocgdc"]
+) as dag:
+    submitSparkJobOperator = DataprocGDCSubmitSparkJobKrmOperator(
+        task_id="example-dataprocgdc-submitspark-operator",
+        trigger_rule="all_success",
+        depends_on_past=False,
+        retries=1,
+        application_file="sparkpi.yaml",
+        namespace="default",
+        kubernetes_conn_id="myk8s",
```
Comment (MaksYermak) [inline tests/system/providers/google/gdc/dataprocgdc/example_dataprocgdc.py:44]: Does this operator work with GKE cluster configuration?

File: tests/system/providers/google/gdc/dataprocgdc/example_dataprocgdc.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+from airflow import DAG
+from airflow.providers.google.gdc.operators.dataprocgdc import (
+    DataprocGdcCreateAppEnvironmentKrmOperator,
+    DataprocGDCSubmitSparkJobKrmOperator,
+)
+from airflow.providers.google.gdc.sensors.dataprocgdc import DataprocGDCKrmSensor
+
+DAG_ID = "example_dag_dpgdc-krm"
+
+with DAG(
```
Comment (MaksYermak) [inline tests/system/providers/google/gdc/dataprocgdc/example_dataprocgdc.py:36]: Could you please update this system test using project codestyle e.g. https://github.com/apache/airflow/blob/main/tests/system/providers/google/cloud/kubernetes_engine/example_kubernetes_engine.py ?

**Review by MaksYermak — CHANGES_REQUESTED** [review]

File: airflow/providers/google/cloud/operators/dataprocgdc.py (hunk 2 above)
Comment (MaksYermak) [reply airflow/providers/google/cloud/operators/dataprocgdc.py]: @akashsriv07 could you please answer my question above?

File: tests/system/providers/google/gdc/dataprocgdc/example_dataprocgdc.py (hunk 7 above)
Comment (michalmodras) [reply tests/system/providers/google/gdc/dataprocgdc/example_dataprocgdc.py:36]: +1

**Comment by eladkal** [comment]
> Hey, Michal from Cloud Composer team here.
+1 for this code living in a separate, GDC provider - GDC is based on slightly different infra than GCP, and the operators won't necessarily work in the same way (some might be very similar, some very different).

Given the above statemant and the fact that we are going to seprate google to several providers we should treat this PR like adding a new provider which means it nreds to go through the approval cycle like all other providers.
https://github.com/apache/airflow/blob/main/PROVIDERS.rst#accepting-new-community-providers

> If there's shared code, I propose having a util module/provider between these two.

That will not work and this is one of the reason why seperating Google provider is not an easy task. You can't have utill shared between two providers. Utill belongs to one provider. Maybe Google can intoduce google.common provider but that would require separate discussion.
I really advise against adding more completely to Google provider it will make things harder for the decouple project

**Review by eladkal — CHANGES_REQUESTED** [review]
Setting request changes to avoid accedental merge

**Comment by akashsriv07** [comment]
> Setting request changes to avoid accedental merge

Will explore the feasibility on creating the provider and the process to be followed. Thanks

**Pull request closed without merging**